asterius/rts/rts.scheduler.mjs

```diff
@@ -0,0 +1,475 @@
+import { Channel } from "./rts.channel.mjs";
+import { Memory } from "./rts.memory.mjs";
+import * as rtsConstants from "./rts.constants.mjs";
+
+/**
+ * Scheduler.
+ *
+ * TSO stands for Thread State Object.
+ *
+ * @property tsos     Contains info (tid, addr, status...) about all the TSOs.
+ * @property runQueue Queue of runnable TSOs.
+ *
+ */
+export class Scheduler {
+  constructor(memory, symbol_table, stablePtrManager) {
+    this.channel = new Channel(),
+    this.memory = memory;
+    this.symbolTable = symbol_table;
+    this.lastTid = 0;
+    this.tsos = new Map();         // all the TSOs
+    this.runQueue = [];            // Runnable TSO IDs
+    this.blockedTSOs = new Set();  // Blocked TSO IDs
+    this.completeTSOs = new Set(); // Finished TSO IDs
+    this.exports = undefined;
+    this.stablePtrManager = stablePtrManager;
+    this.gc = undefined;
+    this.running = false;
+    Object.seal(this);
+  }
+
+  setGC(gc) {
+    this.gc = gc;
+  }
+
+  /**
+   * Create a new TSO. Called by "createThread"
+   *
+   * @returns Number TSO ID.
+   */
+  newTSO() {
+    const tid = ++this.lastTid;
+    let promise_res = undefined;
+    let promise_rej = undefined;
+    let p = new Promise( (res,rej) =>
+                  { promise_res = res
+                  ; promise_rej = rej
+                  });
+    this.tsos.set(
+      tid,
+      Object.seal({
+        addr: -1,         // TSO struct address in Wasm memory
+        ret: 0,           // returned object address in Wasm memory
+        rstat: -1,        // thread status
+        ffiAsyncFunc: 0,  // FFI asynchronous func
+        ffiRet:undefined, // FFI returned value
+        ffiRetType:0,     // FFI returned value type
+        ffiRetErr:undefined, // FFI returned error
+        promise: p,       // Promise used by user code to wait on this thread
+        promise_resolve: promise_res,
+        promise_reject:  promise_rej,
+        blockingPromise: undefined // Promise used to block on JS FFI code
+      })
+    );
+    return tid;
+  }
+
+  getTSOaddr(tid) {
+    return this.tsos.get(tid).addr;
+  }
+
+  getTSOret(tid) {
+    return this.tsos.get(tid).ret;
+  }
+
+  getTSOrstat(tid) {
+    return this.tsos.get(tid).rstat;
+  }
+
+  setTSOaddr(tid, addr) {
+    this.tsos.get(tid).addr = addr;
+  }
+
+  setTSOret(tid, ret) {
+    this.tsos.get(tid).ret = ret;
+  }
+
+  setTSOrstat(tid, rstat) {
+    this.tsos.get(tid).rstat = rstat;
+  }
+
+  getTSOid(tso) {
+    return this.memory.i32Load(tso + rtsConstants.offset_StgTSO_id);
+  }
+
+  /**
+   * Called from a generated safe FFI import call.
+   *
+   * @param ffiPromise Promise executing the FFI import code asynchronously.
+   */
+  returnFFIPromise(tid, ffiPromise) {
+    const tso_info = this.tsos.get(tid);
+    //console.log(`Thread ${tid}: return FFI Promise`);
+    tso_info.blockingPromise = ffiPromise;
+  }
+
+  /**
+   * Called when a thread stops for some reason.
+   */
+  returnFromTSO(tid) {
```

Comment:
Minor naming suggestion: `returnedFromTSO` is nicer since it's called *after* another round of cmm execution, but `returnFromTSO` sounds like it's used *during* execution to terminate it.